src/main/scala/com/databricks/spark/redshift/RedshiftJDBCWrapper.scala

```diff
@@ -194,9 +199,19 @@ private[redshift] class JDBCWrapper {
    */
   def getConnector(userProvidedDriverClass: Option[String], url: String): Connection = {
     val subprotocol = url.stripPrefix("jdbc:").split(":")(0)
-    val driverClass: Class[Driver] = getDriverClass(subprotocol, userProvidedDriverClass)
-    registerDriver(driverClass.getCanonicalName)
-    DriverManager.getConnection(url, new Properties())
+    val driverClass: String = getDriverClass(subprotocol, userProvidedDriverClass)
+    registerDriver(driverClass)
+    // Note that we purposely don't call DriverManager.getConnection() here: we want to ensure
+    // that an explicitly-specified user-provided driver class can take precedence over the default
+    // class, but DriverManager.getConnection() might return a according to a different precedence.
+    // At the same time, we don't want to create a driver-per-connection, so we use the
+    // DriverManager's driver instances to handle that singleton logic for us.
+    val driver: Driver = DriverManager.getDrivers.asScala.collectFirst {
+      case d if d.getClass.getCanonicalName == driverClass => d
```

Comment:
I just realized that there's a bit of a bug here: if `d` is an instance of `DriverWrapper` then we need to compare against `d.wrapped.getClass.getCanonicalName`. I'll open a followup PR to fix this.